core/opaevaluator.go

```diff
@@ -365,65 +378,19 @@ func (evaluator *OPAEvaluator) PolicyEvaluation(logger *logrus.Entry, permission
 	return dataFromEvaluation, nil, nil
 }
 
-func CreateRegoQueryInput(req *http.Request, env config.EnvironmentVariables, enableResourcePermissionsMapOptimization bool, user types.User, responseBody interface{}) ([]byte, error) {
-	requestContext := req.Context()
-	logger := glogger.Get(requestContext)
-	opaInputCreationTime := time.Now()
-	userProperties := make(map[string]interface{})
-	_, err := utils.UnmarshalHeader(req.Header, env.UserPropertiesHeader, &userProperties)
-	if err != nil {
-		return nil, fmt.Errorf("user properties header is not valid: %s", err.Error())
-	}
-
-	userGroup := make([]string, 0)
-	userGroupsNotSplitted := req.Header.Get(env.UserGroupsHeader)
-	if userGroupsNotSplitted != "" {
-		userGroup = strings.Split(userGroupsNotSplitted, ",")
-	}
-
-	var permissionsMap PermissionsOnResourceMap
-	if enableResourcePermissionsMapOptimization {
-		logger.Info("preparing optimized resourcePermissionMap for OPA evaluator")
-		opaPermissionsMapTime := time.Now()
-		permissionsMap = buildOptimizedResourcePermissionsMap(user)
-		logger.WithField("resourcePermissionMapCreationTime", fmt.Sprintf("%+v", time.Since(opaPermissionsMapTime))).Tracef("resource permission map creation")
-	}
+type RegoInputOptions struct {
+	EnableResourcePermissionsMapOptimization bool
+}
 
-	input := Input{
-		ClientType: req.Header.Get(env.ClientTypeHeader),
-		Request: InputRequest{
-			Method:     req.Method,
-			Path:       req.URL.Path,
-			Headers:    req.Header,
-			Query:      req.URL.Query(),
-			PathParams: mux.Vars(req),
-		},
-		Response: InputResponse{
-			Body: responseBody,
-		},
-		User: InputUser{
-			Bindings:               user.UserBindings,
-			Roles:                  user.UserRoles,
-			Properties:             userProperties,
-			Groups:                 userGroup,
-			ResourcePermissionsMap: permissionsMap,
-		},
-	}
+func CreateRegoQueryInput(
+	logger *logrus.Entry,
+	input Input,
```

Comment:
I'm not convinced about separating input creation from the rego input creation; they are the same and I'd require the SDK user to invoke two functions rather than one. 

Your idea was to make InputFromRequest interchangeable with other InputFromFiberRequest InputFromSomethingElse, etc?

If that's the case I would not create different functions but create a single function that accepts an interface and if you want to build the input from something else you have to implement the interface for that (rather than duplicating the input creation). This way I'd stick with a single function `CreateRegoQueryInput` (which i'd rename to `CreateRegoInput` or simply `CreateInput`) which accepts an `InputSource` interface (we can provide some ready to use implementations, e.g for `*http.Request`)

Reply:
Yes, my idea is to make InputFromRequest interchangeable.
At the moment, this is the cleaner way to do it as first step to avoid a huge PR. The basic idea is to create the sdk and I'm not sure at the moment if the core package should manage the input creation or it should be done by sdk.

At the moment the scope is to separate the functions to better understand the different ownership.

Reply:
mmmh not yet convinced but ok